File: astrocyte-local-py/astrocyte_local/context_tree.py
"""Context Tree — hierarchical markdown file storage.

Stores memories as .md files with YAML frontmatter in a domain-based
directory structure. See docs/context-tree-format.md for the spec.

All operations are sync (file I/O). Async wrappers in engine.py.
"""

from __future__ import annotations

import re
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import yaml


@dataclass
class MemoryEntry:
    """A single memory stored as a markdown file."""

    id: str
    bank_id: str
    text: str
    domain: str
    file_path: str  # Relative to memory root
    memory_layer: str = "fact"  # "fact", "observation", "model"
    fact_type: str = "world"  # "world", "experience", "observation"
    tags: list[str] = field(default_factory=list)
    created_at: str = ""  # ISO 8601
    updated_at: str = ""
    occurred_at: str | None = None
    recall_count: int = 0
    last_recalled_at: str | None = None
    source: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


class ContextTree:
    """Hierarchical markdown file storage for memories.

    Directory structure:
        {root}/memory/{domain}/{filename}.md

    Each .md file has YAML frontmatter + plain text content.
    """

    def __init__(self, root: str | Path) -> None:
        self.root = Path(root)
        self.memory_dir = self.root / "memory"
        self.memory_dir.mkdir(parents=True, exist_ok=True)

    def store(
        self,
        content: str,
        bank_id: str,
        *,
        domain: str = "general",
        tags: list[str] | None = None,
        memory_layer: str = "fact",
        fact_type: str = "world",
        occurred_at: str | None = None,
        source: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> MemoryEntry:
        """Store a new memory as a markdown file. Returns the created entry."""
        mem_id = uuid.uuid4().hex[:12]
        now = datetime.now(timezone.utc).isoformat()
        filename = self._make_filename(content)

        # Ensure domain directory exists
        domain_dir = self.memory_dir / domain
        domain_dir.mkdir(parents=True, exist_ok=True)

        # Handle filename collisions
        file_path = domain_dir / f"{filename}.md"
        counter = 2
        while file_path.exists():
            file_path = domain_dir / f"{filename}-{counter}.md"
            counter += 1

        rel_path = str(file_path.relative_to(self.memory_dir))

        entry = MemoryEntry(
            id=mem_id,
            bank_id=bank_id,
            text=content,
            domain=domain,
            file_path=rel_path,
            memory_layer=memory_layer,
            fact_type=fact_type,
            tags=tags or [],
            created_at=now,
            updated_at=now,
            occurred_at=occurred_at,
            source=source,
            metadata=metadata or {},
        )

        self._write_entry(file_path, entry)
        return entry

    def read(self, entry_id: str) -> MemoryEntry | None:
        """Read a memory by ID. Scans all files. Returns None if not found."""
        for entry in self.scan_all():
            if entry.id == entry_id:
                return entry
        return None

    def update(self, entry_id: str, content: str) -> MemoryEntry | None:
        """Update a memory's content. Returns updated entry or None."""
        for md_file in self.memory_dir.rglob("*.md"):
            entry = self._read_file(md_file)
            if entry and entry.id == entry_id:
                entry.text = content
                entry.updated_at = datetime.now(timezone.utc).isoformat()
                self._write_entry(md_file, entry)
                return entry
        return None

    def delete(self, entry_id: str) -> bool:
        """Delete a memory by ID. Returns True if found and deleted."""
        for md_file in self.memory_dir.rglob("*.md"):
            entry = self._read_file(md_file)
            if entry and entry.id == entry_id:
                md_file.unlink()
                # Remove empty domain directories
                parent = md_file.parent
                if parent != self.memory_dir and not any(parent.iterdir()):
                    parent.rmdir()
                return True
        return False

    def record_recall(self, entry_id: str) -> None:
        """Increment recall_count and update last_recalled_at for an entry."""
        for md_file in self.memory_dir.rglob("*.md"):
            entry = self._read_file(md_file)
            if entry and entry.id == entry_id:
                entry.recall_count += 1
                entry.last_recalled_at = datetime.now(timezone.utc).isoformat()
                self._write_entry(md_file, entry)
                return

    def list_domains(self, bank_id: str | None = None) -> list[str]:
        """List all domain directories."""
        domains: list[str] = []
        if not self.memory_dir.exists():
            return domains
        for d in sorted(self.memory_dir.iterdir()):
            if d.is_dir() and not d.name.startswith("_"):
                if bank_id is None:
                    domains.append(d.name)
                else:
                    # Check if any file in this domain matches the bank
                    for md in d.glob("*.md"):
                        entry = self._read_file(md)
                        if entry and entry.bank_id == bank_id:
                            domains.append(d.name)
                            break
        return domains

    def list_entries(self, bank_id: str, domain: str | None = None) -> list[MemoryEntry]:
        """List entries in a domain (or all domains) for a bank."""
        entries: list[MemoryEntry] = []
        search_dir = self.memory_dir / domain if domain else self.memory_dir
        if not search_dir.exists():
            return entries
        for md_file in sorted(search_dir.rglob("*.md")):
            entry = self._read_file(md_file)
            if entry and entry.bank_id == bank_id:
                entries.append(entry)
        return entries

    def scan_all(self, bank_id: str | None = None) -> list[MemoryEntry]:
        """Scan all memory files. Optionally filter by bank_id."""
        entries: list[MemoryEntry] = []
        if not self.memory_dir.exists():
            return entries
        for md_file in sorted(self.memory_dir.rglob("*.md")):
            entry = self._read_file(md_file)
            if entry:
                if bank_id is None or entry.bank_id == bank_id:
                    entries.append(entry)
        return entries

    def count(self, bank_id: str | None = None) -> int:
        """Count total memories."""
        return len(self.scan_all(bank_id))

    # ── Internal ──

    def _make_filename(self, content: str) -> str:
        """Generate a filename from content (first 50 chars, slugified)."""
        slug = content[:50].lower().strip()
        slug = re.sub(r"[^a-z0-9]+", "-", slug)
        slug = slug.strip("-")
        return slug or "memory"

    def _write_entry(self, path: Path, entry: MemoryEntry) -> None:
        """Write a MemoryEntry as a markdown file with YAML frontmatter."""
        frontmatter: dict[str, Any] = {
            "id": entry.id,
            "bank_id": entry.bank_id,
            "memory_layer": entry.memory_layer,
            "fact_type": entry.fact_type,
            "tags": entry.tags,
            "created_at": entry.created_at,
            "updated_at": entry.updated_at,
            "recall_count": entry.recall_count,
        }
        if entry.occurred_at:
            frontmatter["occurred_at"] = entry.occurred_at
        if entry.last_recalled_at:
            frontmatter["last_recalled_at"] = entry.last_recalled_at
        if entry.source:
            frontmatter["source"] = entry.source
        if entry.metadata:
            frontmatter["metadata"] = entry.metadata

        fm_str = yaml.dump(frontmatter, default_flow_style=False, sort_keys=False)
        path.write_text(f"---\n{fm_str}---\n\n{entry.text}\n", encoding="utf-8")

    def _read_file(self, path: Path) -> MemoryEntry | None:
        """Read a markdown file and parse into MemoryEntry."""
        try:
            content = path.read_text(encoding="utf-8")
        except (OSError, UnicodeDecodeError):
            return None

        # Parse YAML frontmatter
        if not content.startswith("---"):
            return None

        parts = content.split("---", 2)
        if len(parts) < 3:
            return None

        try:
            fm = yaml.safe_load(parts[1]) or {}
        except yaml.YAMLError:
            return None

        text = parts[2].strip()
        rel_path = str(path.relative_to(self.memory_dir))
        domain = path.parent.name if path.parent != self.memory_dir else "general"

        return MemoryEntry(
            id=fm.get("id", ""),
            bank_id=fm.get("bank_id", ""),
            text=text,
            domain=domain,
            file_path=rel_path,
            memory_layer=fm.get("memory_layer", "fact"),
            fact_type=fm.get("fact_type", "world"),
            tags=fm.get("tags", []),
            created_at=fm.get("created_at", ""),
            updated_at=fm.get("updated_at", ""),
            occurred_at=fm.get("occurred_at"),
            recall_count=fm.get("recall_count", 0),
            last_recalled_at=fm.get("last_recalled_at"),
            source=fm.get("source"),
            metadata=fm.get("metadata", {}),
        )

File: astrocyte-local-ts/src/context-tree.ts
/**
 * Context Tree — hierarchical markdown file storage.
 *
 * Stores memories as .md files with YAML frontmatter in a
 * domain-based directory structure. See docs/context-tree-format.md.
 */

import fs from "node:fs";
import path from "node:path";
import { randomUUID } from "node:crypto";
import YAML from "yaml";
import type { MemoryEntry } from "./types.js";

export class ContextTree {
  private memoryDir: string;

  constructor(private root: string) {
    this.memoryDir = path.join(root, "memory");
    fs.mkdirSync(this.memoryDir, { recursive: true });
  }

  store(options: {
    content: string;
    bank_id: string;
    domain?: string;
    tags?: string[];
    memory_layer?: "fact" | "observation" | "model";
    fact_type?: "world" | "experience" | "observation";
    occurred_at?: string;
    source?: string;
    metadata?: Record<string, string | number | boolean | null>;
  }): MemoryEntry {
    const id = randomUUID().replace(/-/g, "").slice(0, 12);
    const now = new Date().toISOString();
    const domain = options.domain || "general";
    const domainDir = path.join(this.memoryDir, domain);
    fs.mkdirSync(domainDir, { recursive: true });

    const filename = this.makeFilename(options.content);
    let filePath = path.join(domainDir, `${filename}.md`);
    let counter = 2;
    while (fs.existsSync(filePath)) {
      filePath = path.join(domainDir, `${filename}-${counter}.md`);
      counter++;
    }

    const relPath = path.relative(this.memoryDir, filePath);

    const entry: MemoryEntry = {
      id,
      bank_id: options.bank_id,
      text: options.content,
      domain,
      file_path: relPath,
      memory_layer: options.memory_layer || "fact",
      fact_type: options.fact_type || "world",
      tags: options.tags || [],
      created_at: now,
      updated_at: now,
      occurred_at: options.occurred_at,
      recall_count: 0,
      source: options.source,
      metadata: options.metadata || {},
    };

    this.writeEntry(filePath, entry);
    return entry;
  }

  read(entryId: string): MemoryEntry | null {
    for (const entry of this.scanAll()) {
      if (entry.id === entryId) return entry;
    }
    return null;
  }

  update(entryId: string, content: string): MemoryEntry | null {
    for (const mdFile of this.allMdFiles()) {
      const entry = this.readFile(mdFile);
      if (entry && entry.id === entryId) {
        entry.text = content;
        entry.updated_at = new Date().toISOString();
        this.writeEntry(mdFile, entry);
        return entry;
      }
    }
    return null;
  }

  delete(entryId: string): boolean {
    for (const mdFile of this.allMdFiles()) {
      const entry = this.readFile(mdFile);
      if (entry && entry.id === entryId) {
        fs.unlinkSync(mdFile);
        // Remove empty domain directories
        const parent = path.dirname(mdFile);
        if (parent !== this.memoryDir) {
          try {
            const remaining = fs.readdirSync(parent);
            if (remaining.length === 0) fs.rmdirSync(parent);
          } catch {}
        }
        return true;
      }
    }
    return false;
  }

  recordRecall(entryId: string): void {
    for (const mdFile of this.allMdFiles()) {
      const entry = this.readFile(mdFile);
      if (entry && entry.id === entryId) {
        entry.recall_count++;
        entry.last_recalled_at = new Date().toISOString();
        this.writeEntry(mdFile, entry);
        return;
      }
    }
  }

  listDomains(bankId?: string): string[] {
    if (!fs.existsSync(this.memoryDir)) return [];
    const domains: string[] = [];
    for (const d of fs.readdirSync(this.memoryDir, { withFileTypes: true }).sort((a, b) => a.name.localeCompare(b.name))) {
      if (!d.isDirectory() || d.name.startsWith("_")) continue;
      if (!bankId) {
        domains.push(d.name);
      } else {
        const domainPath = path.join(this.memoryDir, d.name);
        const files = fs.readdirSync(domainPath).filter((f) => f.endsWith(".md"));
        for (const f of files) {
          const entry = this.readFile(path.join(domainPath, f));
          if (entry && entry.bank_id === bankId) {
            domains.push(d.name);
            break;
          }
        }
      }
    }
    return domains;
  }

  listEntries(bankId: string, domain?: string): MemoryEntry[] {
    const searchDir = domain ? path.join(this.memoryDir, domain) : this.memoryDir;
    if (!fs.existsSync(searchDir)) return [];
    const entries: MemoryEntry[] = [];
    for (const mdFile of this.allMdFiles(searchDir)) {
      const entry = this.readFile(mdFile);
      if (entry && entry.bank_id === bankId) entries.push(entry);
    }
    return entries;
  }

  scanAll(bankId?: string): MemoryEntry[] {
    if (!fs.existsSync(this.memoryDir)) return [];
    const entries: MemoryEntry[] = [];
    for (const mdFile of this.allMdFiles()) {
      const entry = this.readFile(mdFile);
      if (entry && (!bankId || entry.bank_id === bankId)) entries.push(entry);
    }
    return entries;
  }

  count(bankId?: string): number {
    return this.scanAll(bankId).length;
  }

  // ── Internal ──

  private makeFilename(content: string): string {
    let slug = content.slice(0, 50).toLowerCase().trim();
    slug = slug.replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "");
    return slug || "memory";
  }

  private writeEntry(filePath: string, entry: MemoryEntry): void {
    const frontmatter: Record<string, unknown> = {
      id: entry.id,
      bank_id: entry.bank_id,
      memory_layer: entry.memory_layer,
      fact_type: entry.fact_type,
      tags: entry.tags,
      created_at: entry.created_at,
      updated_at: entry.updated_at,
      recall_count: entry.recall_count,
    };
    if (entry.occurred_at) frontmatter.occurred_at = entry.occurred_at;
    if (entry.last_recalled_at) frontmatter.last_recalled_at = entry.last_recalled_at;
    if (entry.source) frontmatter.source = entry.source;
    if (Object.keys(entry.metadata).length > 0) frontmatter.metadata = entry.metadata;

    const fmStr = YAML.stringify(frontmatter);
    fs.writeFileSync(filePath, `---\n${fmStr}---\n\n${entry.text}\n`, "utf-8");
  }

  private readFile(filePath: string): MemoryEntry | null {
    try {
      const content = fs.readFileSync(filePath, "utf-8");
      if (!content.startsWith("---")) return null;

      const parts = content.split("---");
      if (parts.length < 3) return null;

      const fm = YAML.parse(parts[1]) || {};
      const text = parts.slice(2).join("---").trim();
      const relPath = path.relative(this.memoryDir, filePath);
      const domain = path.dirname(filePath) === this.memoryDir ? "general" : path.basename(path.dirname(filePath));

      return {
        id: fm.id || "",
        bank_id: fm.bank_id || "",
        text,
        domain,
        file_path: relPath,
        memory_layer: fm.memory_layer || "fact",
        fact_type: fm.fact_type || "world",
        tags: fm.tags || [],
        created_at: fm.created_at || "",
        updated_at: fm.updated_at || "",
        occurred_at: fm.occurred_at,
        recall_count: fm.recall_count || 0,
        last_recalled_at: fm.last_recalled_at,
        source: fm.source,
        metadata: fm.metadata || {},
      };
    } catch {
      return null;
    }
  }

  private *allMdFiles(dir?: string): Generator<string> {
    const searchDir = dir || this.memoryDir;
    if (!fs.existsSync(searchDir)) return;
    for (const entry of fs.readdirSync(searchDir, { withFileTypes: true })) {
      const fullPath = path.join(searchDir, entry.name);
      if (entry.isDirectory() && !entry.name.startsWith("_")) {
        yield* this.allMdFiles(fullPath);
      } else if (entry.isFile() && entry.name.endsWith(".md")) {
        yield fullPath;
      }
    }
  }
}
